lib/mainloop.c

```diff
@@ -365,6 +366,51 @@ main_loop_verify_config(GString *result, MainLoop *self)
   g_free(file_contents);
 }
 
+/* main_loop_config_monitor_init
+ * initializes configuration monitoring */
+
+/* callback */
+static void
+_on_file_modified(const DirectoryMonitorEvent *event, gpointer user_data)
+{
+  MainLoop *self = (MainLoop *) user_data;
+
+  if (!strcmp(event->full_path, resolvedConfigurablePaths.cfgfilename)
+      && ((event->event_type == FILE_MODIFIED) || (event->event_type == FILE_CREATED))
+      && (monitor->initial_collect == FALSE))
```

Comment:
I do not like the idea of `initial_collect` flag, but I guess this is triggered for the `FILE_CREATED` when it starts.
* I would either say to make the directory monitor configurable whether it should ignore the initial triggers at first scan or not. (this way it is hidden in the directory monitor)
* Handle here that it ignores the first `FILE_CREATED` for the `cfgfilename` (we can assume if we are here that there was a file when syslog-ng started).

I guess it is more like the call side responsibility, so I would do the second, even thou that not pretty.
Something like this:
```c
static void
 _on_file_modified(const DirectoryMonitorEvent *event, gpointer user_data)
 {
   static gboolean ignore_first_time_created_event = TRUE;
   MainLoop *self = (MainLoop *) user_data;

   if (strcmp(event->full_path, resolvedConfigurablePaths.cfgfilename))
      return;

   if (ignore_first_time_created_event && event_type == FILE_CREATED)
     {
         ignore_first_time_created_event = FALSE;
         return;
     } 

   if ((event->event_type == FILE_MODIFIED) || (event->event_type == FILE_CREATED))
     {
       main_loop_reload_config(self);
     }
 }
```
(disclaimer: this is just a sketch and not tested/compiled)

Reply:
Shouldn't ``static gboolean ignore_first_time_created_event = TRUE;`` be outside the callback though?

Reply:
Why should it ? It is only needed in the that function.

Reply:
I'm thinking, it will be set to ``TRUE`` every time the callback is invoked, isn't it?

Reply:
I'm thinking, it will be set to ``TRUE`` every time the callback is invoked, won't it?

Reply:
No, it is assigned only the first time. You can try:
```c#include <stdio.h>

void sayhello()
{
  static int once = 1;

  if (!once) return;

  once = 0;
  printf("hello\n");
}


int main()
{
  sayhello();
  sayhello();
  sayhello();
  sayhello();
 
  return 0;
}

```